### PR TITLE
Support user context

### PR DESCRIPTION
**Pull request by dustin-graham**
allow the client app to specify a user context for logging events and exceptions to sentry.

### CODE REVIEW

**Comment by dustin-graham** [comment]
I signed it!

**Comment by dustin-graham** [comment]
@yjbanov it's failing on the `GZIP` warning. in local testing with my own app I found that when I upgraded this package to use `gzip` instead, it failed at runtime in Flutter because it couldn't be found. seems like a Dart version mismatch maybe, but that's why it's failing Travis right now.

**Comment by dustin-graham** [comment]
@zoechi was this what you where looking for here? 
https://forum.sentry.io/t/usercontext-without-explicit-support-in-the-client-package/2760

File: lib/sentry.dart [hunk 1]
```diff
@@ -389,9 +403,45 @@ class Event {
...
+class User {
+  final String id;
+  final String username;
+  final String email;
+  final String ipAddress;
+  final String subscription;
+
+  const User(this.id, this.username, this.email, this.ipAddress, this.subscription);
+
+  Map<String, String> toJson() {
```
Comment (yjbanov) [inline lib/sentry.dart]: dartdocs

**Review by yjbanov — COMMENTED** [review]
Thanks for the contribution! I most have nitpicks. Design-wise, I'm only curious about making `userContext` mutable.

**Comment by dustin-graham** [comment]
@yjbanov Thanks for the review. I'll push an update tomorrow with those fixes. About the mutable `userContext`. For my application some of the data for the `User` is not known until later on at runtime after the user has logged in and done some other things. Allowing the client application to change the user context at any time allows for this flexibility without having to go make a new SentryClient. In one of my commits I had it the other way, where the user passes in the `userContext` as part of the initial `Event` environment, but I quickly found that that didn't work for me as the data for the context is only known after login and perhaps further initialization. I did leave in the ability to override the base `userContext` when sending individual events in case this needs to be overridden for some reason. I could also see removing the `userContext` from the SentryClient entirely and just make adding in user data an exercise for the user each time they log. But this would likely be burdensome on the client.

**Comment by yjbanov** [comment]
Ok, I think it makes sense. I suppose in some apps you can switch users dynamically too. So let's keep it mutable. Why don't we just make it a mutable field then? The Dart style guide discourages setters without getters.

**Comment by yjbanov** [comment]
Please ping this PR again when you have the changes finalized and I'll look into the GZIP issue. Recently the Dart standard libraries deprecated ALL_CAPS constant names in favor of camelCase.

File: lib/sentry.dart (hunk 1 above)
Comment (dustin-graham) [reply lib/sentry.dart]: done

File: lib/sentry.dart [hunk 2]
```diff
@@ -389,9 +403,45 @@ class Event {
...
+///}
+///
+class User {
+  final String id;
+  final String username;
+  final String email;
+  final String ipAddress;
+  final String subscription;
+
+  const User(this.id, this.username, this.email, this.ipAddress, this.subscription);
```
[the remark this answers is not in the document]
Comment (dustin-graham) [reply lib/sentry.dart]: none of them are non-null. The docs say that you must have at least an ID OR an IP Address set. but that is up to the client app to enforce which they want. adding an assert to inform the developer during debug that there might be an issue.

File: lib/sentry.dart [hunk 3]
```diff
@@ -389,9 +403,45 @@ class Event {
...
+///  "ip_address": "127.0.0.1",
+///  "subscription": "basic"
+///}
+///
+class User {
+  final String id;
+  final String username;
+  final String email;
+  final String ipAddress;
+  final String subscription;
```
[the remark this answers is not in the document]
Comment (dustin-graham) [reply lib/sentry.dart]: done

File: lib/sentry.dart [hunk 4]
```diff
@@ -389,9 +403,45 @@ class Event {
...
+/// Conforms to the User Interface contract for Sentry
+/// https://docs.sentry.io/clientdev/interfaces/user/
+///
+/// "user": {
+///  "id": "unique_id",
+///  "username": "my_user",
+///  "email": "foo@example.com",
+///  "ip_address": "127.0.0.1",
+///  "subscription": "basic"
+///}
```
[the remark this answers is not in the document]
Comment (dustin-graham) [reply lib/sentry.dart]: done

File: lib/sentry.dart [hunk 5]
```diff
@@ -330,6 +342,8 @@ class Event {
   /// they must be JSON-serializable.
   final Map<String, dynamic> extra;
 
+  final User userContext;
```
[the remark this answers is not in the document]
Comment (dustin-graham) [reply lib/sentry.dart]: done

File: lib/sentry.dart [hunk 6]
```diff
@@ -143,6 +143,13 @@ class SentryClient {
   /// Attached to the event payload.
   final String projectId;
 
+  @visibleForTesting
+  User _userContext;
+
+  set userContext(User val) {
```
[the remark this answers is not in the document]
Comment (dustin-graham) [reply lib/sentry.dart]: done

**Comment by dustin-graham** [comment]
@yjbanov I believe I have addressed all the feedback and Travis is now passing.

File: test/sentry_test.dart [hunk 7]
```diff
@@ -74,7 +74,9 @@ void main() {
         'Content-Type': 'application/json',
         'X-Sentry-Auth': 'Sentry sentry_version=6, '
             'sentry_client=${SentryClient.sentryClient}, '
-            'sentry_timestamp=${fakeClock.now().millisecondsSinceEpoch}, '
+            'sentry_timestamp=${fakeClock
+            .now()
+            .millisecondsSinceEpoch}, '
```
Comment (yjbanov) [inline test/sentry_test.dart:79]: Wow, did `dartfmt` do this?

Comment (dustin-graham) [reply test/sentry_test.dart:79]: it did. it appears that the travis rule runs dartfmt and fails if there is anything changed after running it. so after I ran dartfmt locally and uploaded it now passes.

**Review by yjbanov — APPROVED** [review]

File: test/sentry_test.dart (hunk 7 above)
Comment (yjbanov) [reply test/sentry_test.dart:79]: I should file a bug.

**Comment by yjbanov** [comment]
lgtm

Can I merge this? I'll also do a release.

**Pull request merged**